pkg/services/govmomi/vcenter/clone.go

```diff
@@ -390,6 +404,105 @@ func getDiskConfigSpec(disk *types.VirtualDisk, diskCloneCapacityKB int64) (type
 	}, nil
 }
 
+// createDataDisks parses through the list of VSphereDisk objects and generates the VirtualDeviceConfigSpec for each disk needing to be created during clone process.
+func createDataDisks(ctx context.Context, disks []infrav1.VSphereDisk, primaryDisk *types.VirtualDisk, devices object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	log := ctrl.LoggerFrom(ctx)
+	additionalDisks := []types.BaseVirtualDeviceConfigSpec{}
+	unit := int32(1)
+
+	for i, dataDisk := range disks {
+		log.Info("Adding disk", "spec", dataDisk)
+
+		//  Get controller.  Only supporting using same controller as primary disk at this time
+		controller, ok := devices.FindByKey(primaryDisk.ControllerKey).(types.BaseVirtualController)
+		if !ok {
+			return nil, errors.Errorf("unable to find controller with key=%v", primaryDisk.ControllerKey)
+		}
+
+		dev := &types.VirtualDisk{
+			VirtualDevice: types.VirtualDevice{
+				Key: devices.NewKey() - int32(i),
+				Backing: &types.VirtualDiskFlatVer2BackingInfo{
+					DiskMode:        string(types.VirtualDiskModePersistent),
+					ThinProvisioned: types.NewBool(true),
+					VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
+						FileName: "",
+					},
+				},
+				ControllerKey: controller.GetVirtualController().Key,
+			},
+			CapacityInKB: int64(dataDisk.SizeGiB) * 1024 * 1024,
+		}
+
+		// Assign unit number to next slot on controller
+		assignUnitNumber(ctx, dev, devices, additionalDisks, controller, unit)
+		unit = *dev.UnitNumber
+
+		diskConfigSpec := types.VirtualDeviceConfigSpec{
+			Device:        dev,
+			Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+			FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+		}
+
+		log.Info("Generated device", "dev", dev)
+
+		additionalDisks = append(additionalDisks, &diskConfigSpec)
+	}
+
+	return additionalDisks, nil
+}
+
+// assignUnitNumber assigns a controller unit number to a device.
+func assignUnitNumber(ctx context.Context, device types.BaseVirtualDevice, existingDevices object.VirtualDeviceList, newDevices []types.BaseVirtualDeviceConfigSpec, controller types.BaseVirtualController, offset int32) {
+	log := ctrl.LoggerFrom(ctx)
+	vd := device.GetVirtualDevice()
+	vd.ControllerKey = controller.GetVirtualController().Key
+	vd.UnitNumber = &offset
+
+	units := make([]bool, 30)
+
+	for i := 0; i < int(offset); i++ {
```

Comment:
what is the significance of 30 here?
offset is of type int32, isn't doing len(offset) returning an error here?
> invalid argument: offset (variable of type int32) for built-in len

Reply:
they are doing this though:
```
	units := make([]bool, 30)

	for i := 0; i < offset; i++ {
		units[i] = true
	}
```

i think `len(offset)` would trigger a compiler error.

Reply:
I think what we are wanting to do here is two things.

1.) create an array that is the max number of units any controller can hold (or maybe the one we are targeting, but not sure how to figure that).  I think I read that vSphere for some controllers max is 16, but I think special cases are much much higher.  We can adjust what the default array size is here to the max of any, but will leave future issues popping up where we now exceeded that controller.
2.) The loop using offset is just gonna prefill the array up to the offset point as true as to prevent us from assigning an index below that for the current disk we are attempting to add / create.  This loop seems to match both govmomi and this function.  The govmomi is doing len(offset) because their function is using int32 as type for offset, but mine is just int.

The main function that calls this one, we default the starting point to 0 (which should always be the primary disk).  After that, its any additional disks in the vSphere OVA template or previously added data disks.  If we have more disks being added then supported, I do want to make sure that something triggers a failure.  We could add a check here to error out if offset is > 30 and state it exceeds the max unit number.

I am happy for any suggestions.  This is probably a weak spot, but for the average user of adding 1-5 extra disks (i'm just guessing 5 is max we want to add for each node for OS usage that is not included w/ any of the PVs) I think 30 is an ok number to default with initially.  What do you think?

Reply:
> create an array that is the max number of units any controller can hold 

aren't there exported constants for that? would the govmomi maintainers know?

i think adding a few comment in the code to clarify the choices of 30 and the rest of the logic will clarify the picture for readers. but we should avoid panics if arrays are accessed out of bounds.

my main point was that the code as is cannot work. please try this:
https://go.dev/play/p/fWud2MEYR0d

Reply:
i think docs.vmware.com might be going down in a few days due to broadcom docs migration, but you can still link that URL and a quote from it. if someone needs the original page they can use web archived version.

Reply:
ah good point.  I'll see if I can find the broadcom equivelent.  I keep forgetting that things are in flux with these sites.